core/formats/formats_burst_trie.cpp

```diff
@@ -225,16 +274,116 @@ struct block_t : private util::noncopyable {
-};                      // block_t
+};
+
+template<typename T>
+class MonotonicBuffer {
+  static constexpr size_t kAlign =
+    (alignof(T) * alignof(void*)) / std::gcd(alignof(T), alignof(void*));
+
+  struct alignas(kAlign) Block {
+    Block* prev = nullptr;
+  };
+
+  static_assert(std::is_trivially_destructible_v<Block>);
+
+ public:
+  MonotonicBuffer(IResourceManager& resource_manager,
+                  size_t initial_size) noexcept
+    : resource_manager_{resource_manager}, next_size_{initial_size} {
+    IRS_ASSERT(initial_size > 1);
+  }
+
+  MonotonicBuffer(const MonotonicBuffer&) = delete;
+  MonotonicBuffer& operator=(const MonotonicBuffer&) = delete;
+  ~MonotonicBuffer() { Reset(); }
+
+  template<typename... Args>
+  T* Allocate(Args&&... args) {
+    if (IRS_UNLIKELY(available_ == 0)) {
+      AllocateMemory();
+    }
+    IRS_ASSERT(reinterpret_cast<uintptr_t>(current_) % alignof(T) == 0);
+    auto* p = new (current_) T{std::forward<Args>(args)...};
+    current_ += sizeof(T);
+    --available_;
+    return p;
+  }
+
+  // Release all memory, except current_ and keep next_size_
+  void Clear() noexcept {
+    if (IRS_UNLIKELY(head_ == nullptr)) {
+      return;
+    }
+
+    Release(std::exchange(head_->prev, nullptr));
+    // TODO(MBkkt) Don't be lazy, call Decrease eager
+    // resource_manager_.Decrease(blocks_memory_ - size of head block);
+
+    auto* initial_current = reinterpret_cast<byte_type*>(head_) + sizeof(Block);
+    available_ += (current_ - initial_current) / sizeof(T);
+    current_ = initial_current;
+  }
+
+  // Release all memory, but keep next_size_
+  void Reset() noexcept {
+    if (IRS_UNLIKELY(head_ == nullptr)) {
+      return;
+    }
+
+    Release(std::exchange(head_, nullptr));
+    resource_manager_.Decrease(std::exchange(blocks_memory_, 0));
 
-// FIXME std::is_nothrow_move_constructible_v<std::list<...>> == false
-static_assert(std::is_nothrow_move_constructible_v<block_t>);
-// FIXME std::is_nothrow_move_assignable_v<std::list<...>> == false
-static_assert(std::is_nothrow_move_assignable_v<block_t>);
+    // otherwise we always increasing size!
+    // TODO(MBkkt) we could compute current size, but it's
+    next_size_ = (next_size_ * 2 + 2) / 3;
+
+    available_ = 0;
+    current_ = nullptr;
+  }
+
+ private:
+  void Release(Block* it) noexcept {
+    while (it != nullptr) {
+      operator delete(std::exchange(it, it->prev), std::align_val_t{kAlign});
+    }
+  }
+
+  void AllocateMemory() {
+    const auto size = sizeof(Block) + next_size_ * sizeof(T);
+    // TODO(MBkkt) use allocate_at_least but it's only C++23 :(
+    resource_manager_.Increase(size);
+    blocks_memory_ += size;
+    auto* p =
+      static_cast<byte_type*>(operator new(size, std::align_val_t{kAlign}));
+    head_ = new (p) Block{head_};
+    IRS_ASSERT(p == reinterpret_cast<byte_type*>(head_));
+    current_ = p + sizeof(Block);
+    available_ = next_size_;
+    next_size_ = (next_size_ * 3) / 2;
```

Comment:
We could track first call of `AllocateMemory`   and do increase only if not first. So we will not need tricks with re-counting the size.

Reply:
I don't like if approach because it more cases to think